Remove old actions with no listeners

zbirenbaum opened the pull request:
# Description

Implement scheduler side removal of actions with no listeners. Adds disconnect_timeout_s configuration field with default of 60s. If the client waiting on a given action is disconnected for longer than this duration without reconnecting the scheduler will stop tracking it. This does not remove it from the worker if the job has already been dispatched.

Fixes #338 

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

A test has been added to ensure that requested actions are removed if a client is found to be disconnected for `disconnect_timeout_s`

## Checklist
- [x] Tests added/amended
- [x] `bazel test //...`  passes locally
- [x] PR is contained in a single commit, using `git amend` see some [docs](https://www.atlassian.com/git/tutorials/rewriting-history)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/TraceMachina/nativelink/778)
<!-- Reviewable:end -->

zbirenbaum reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Vercel, pre-commit-checks

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtNs-vl1PJB6EvVwWlS:-NtNs-vmCLvOeFbwGFIJ:b-n7m3m2) ([raw file](https://github.com/TraceMachina/nativelink/blob/4b1c4de69dd76ad88107a7c6c9a87d7160351658/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
> ```XML
> 
>     /// Updated on every client connect and periodically while it has listeners.
>     last_update_timestamp: Mutex<u64>
> ```

Mutex added due to borrow checker difficulties with modifying this field from `active_actions`.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 80 at r1](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtNraCY5GpqXQlG4hOw:-NtNraCY5GpqXQlG4hOx:b8jmgo4) ([raw file](https://github.com/TraceMachina/nativelink/blob/4b1c4de69dd76ad88107a7c6c9a87d7160351658/nativelink-scheduler/src/simple_scheduler.rs#L80)):*
> ```XML
> }
> 
> impl AwaitedAction {
> ```

These helpers are added to avoid in-lining long variable references and mutex locking but might not be necessary


<!-- Sent from Reviewable.io -->

adam-singer reviewed (changes requested):
Should we be using [struct.Instant](https://doc.rust-lang.org/std/time/struct.Instant.html) over [struct.SystemTime](https://doc.rust-lang.org/std/time/struct.SystemTime.html) in this case?

> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778#-:-NtWG5Ga99uu6EpccnFx:b-vnzc15)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), macos-13, ubuntu-22.04, zig-cc ubuntu-20.04, zig-cc ubuntu-22.04

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 475 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWE3012nhQMnPFxd54:-NtWE3012nhQMnPFxd55:b5wahir) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L475)):*
> ```XML
>             // add update to queued action update timestamp here
>             let action = self.queued_actions.get_mut(&action_info).unwrap();
>             let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
> ```

Would it make a difference if we captured system time now before the loop?
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 548 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWDJOt84qynHGY-3Hb:-NtWDJOt84qynHGY-3Hc:b2f56mh) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L548)):*
> ```XML
>         for running_action in running_actions {
>             if running_action.action.notify_channel.receiver_count() > 0 {
>                 running_action.set_last_update_timestamp(SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs());
> ```

grab now before the else/if and having it recalled

```
let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
```


<!-- Sent from Reviewable.io -->

allada reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), macos-13, ubuntu-22.04, zig-cc ubuntu-20.04, zig-cc ubuntu-22.04

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 471 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwQbDM0-6DJULUOcga:-NtwQbDM0-6DJULUOcgb:bwytru4) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L471)):*
> ```XML
>         // unstable feature [see: https://github.com/rust-lang/rust/issues/70530]).
>         let action_infos: Vec<Arc<ActionInfo>> =
>             self.queued_actions.keys().rev().cloned().collect();
> ```

Make this `let queued_actions = self.queued_actions.iter().rev().cloned().collect()`. This will allow you to then use:
```rs
for (action_info, action) in queued_actions {
```

And you won't need to deal with unwrap and such.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwRK51CvxkFx7T_lYg:-NtwRK51CvxkFx7T_lYh:b-r1srxo) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
> ```XML
>             let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
>             if action.notify_channel.receiver_count() > 0 {
>                 action.set_last_update_timestamp(now)
> ```

nit: semi-colon.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwT5As4BIfE0gOh9M_:-NtwT5As4BIfE0gOh9Ma:b-qcxair) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
> ```XML
>             let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
>             if action.notify_channel.receiver_count() > 0 {
>                 action.set_last_update_timestamp(now)
> ```

I really feel that we should be able to detect when clients disconnect and set it then.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 480 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwRgY9AJG2AKt3gQtU:-NtwRgY9AJG2AKt3gQtV:b-34xdy4) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L480)):*
> ```XML
>             } else if action.get_last_update_timestamp() + self.disconnect_timeout_s < now {
>                 self.queued_actions_set.remove(&action_info);
>                 self.queued_actions.remove(&action_info);
> ```

nit: Add a warn message here saying it is being removed.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 545 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwRvmt7WRRdTN4d0MZ:-NtwRvmt7WRRdTN4d0M_:bnw0q68) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L545)):*
> ```XML
> 
>         let mut remove_actions = Vec::new();
>         let running_actions = &mut self.active_actions.values().collect::<Vec<_>>();
> ```

I'm kinda confused on how this ever worked because I believe this is grabbing a reference to a temporary.

In any case you should be beable to just use:

```rs
for running_action in self.active_Actions.values() {
```
since it is not modifying it in the loop.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1634 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSX396ZgxFqnSkeMF:-NtwSX396ZgxFqnSkeMG:b-7kooul) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1634)):*
> ```XML
>         .await?;
> 
>         // Drop our receiver
> ```

nit: period.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1641 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSZQb44xc9y-QHx21:-NtwSZQb44xc9y-QHx22:b-7kooul) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1641)):*
> ```XML
>         tokio::task::yield_now().await;
> 
>         // Sleep for longer than disconnect_timeout_s
> ```

nit: period.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1642 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSoCaDUnpIslmIm58:-NtwSoCaDUnpIslmIm59:bqymbch) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1642)):*
> ```XML
> 
>         // Sleep for longer than disconnect_timeout_s
>         let _ = sleep(Duration::from_secs(DISCONNECT_TIMEOUT_S + 1)).await;
> ```

We should never call native `sleep()` in tests. It's generally considered bad practice. In this case you probably want to hijack the time function inside the scheduler.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1652 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSaK30p7ocNIAEjJV:-NtwSaK30p7ocNIAEjJW:b-7kooul) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1652)):*
> ```XML
>         }
> 
>         // Setup a second action so matching engine is scheduled to rerun
> ```

nit: period.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1665 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSdUSA6u9acwM1urv:-NtwSdUSA6u9acwM1urw:b-7kooul) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1665)):*
> ```XML
>         tokio::task::yield_now().await;
> 
>         // Check to make sure that the action was removed
> ```

nit: period.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1669 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSfmb9BKbbODj8vv8:-NtwSfmb9BKbbODj8vv9:bcpg7mj) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1669)):*
> ```XML
>             .find_existing_action(&unique_qualifier)
>             .await
>             .is_none(),);
> ```

nit: missing message.


<!-- Sent from Reviewable.io -->

allada reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), macos-13, ubuntu-22.04, zig-cc ubuntu-20.04, zig-cc ubuntu-22.04

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwTBf-DF1Q5hn_pZEG:-NtwTBf-DF1Q5hn_pZEH:bfp0zmq) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
> ```XML
> 
>     /// Updated on every client connect and periodically while it has listeners.
>     last_update_timestamp: Mutex<u64>
> ```

I'll help you with this when this PR is closer to being done.

However, `Mutex<(any primitive)>` is silly, since `Atomic{Primitive}` is always better (I can't think of any reason it wouldn't be faster/better, since it's non-blocking and sometimes has hardware support).


<!-- Sent from Reviewable.io -->

zbirenbaum reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Analyze (javascript-typescript), Publish image, Vercel, docker-compose-compiles-nativelink (20.04), pre-commit-checks, ubuntu-20.04 / stable, ubuntu-22.04, vale, and 7 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwTBf-DF1Q5hn_pZEG:-NufJQlm1viKI47VBBLJ:byi55r4) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

I'll help you with this when this PR is closer to being done.

However, `Mutex<(any primitive)>` is silly, since `Atomic{Primitive}` is always better (I can't think of any reason it wouldn't be faster/better, since it's non-blocking and sometimes has hardware support).
</blockquote></details>

I changed it to be AtomicU64 and it works great
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 471 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwQbDM0-6DJULUOcga:-NufJ_3nDCdHt5o9UtjJ:b1zp18x) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L471)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Make this `let queued_actions = self.queued_actions.iter().rev().cloned().collect()`. This will allow you to then use:
```rs
for (action_info, action) in queued_actions {
```

And you won't need to deal with unwrap and such.
</blockquote></details>

`let queued_actions = self.queued_actions.iter().rev().cloned().collect()` gives a type error. I tried a bunch of different things to get it to work but couldn't figure it out.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 475 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWE3012nhQMnPFxd54:-NufK13X5_dvJOVI0VNz:b-kcw19m) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L475)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

Would it make a difference if we captured system time now before the loop?
</blockquote></details>

Do_try_match is the slowest part of the scheduler. If we capture it before the loop and it takes a substantial amount of time to complete, it might fail to time out some actions until the next pass.

Whether or not that it acceptable is something I'm not sure about
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwRK51CvxkFx7T_lYg:-NufKrtN7ECg_MT0E8EY:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: semi-colon.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwT5As4BIfE0gOh9M_:-NufKu6aB3w0Iph4q9q5:beq11pr) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

I really feel that we should be able to detect when clients disconnect and set it then.
</blockquote></details>

I think we discussed this previously and decided that putting this in do_try_match was preferable to creating a new spawn which iterates to check and sleeps. Is there another method of checking when the client drops?
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 480 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwRgY9AJG2AKt3gQtU:-NufT9jtE6Gq8xPrtKeo:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L480)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: Add a warn message here saying it is being removed.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 545 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwRvmt7WRRdTN4d0MZ:-NufTCOsE2TP7YLgD-87:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L545)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

I'm kinda confused on how this ever worked because I believe this is grabbing a reference to a temporary.

In any case you should be beable to just use:

```rs
for running_action in self.active_Actions.values() {
```
since it is not modifying it in the loop.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 548 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWDJOt84qynHGY-3Hb:-NufTjYh-t4kM1gUsvjk:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L548)):*
<details><summary><i>Previously, adam-singer (Adam Singer) wrote…</i></summary><blockquote>

grab now before the else/if and having it recalled

```
let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs();
```
</blockquote></details>

Done.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1634 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSX396ZgxFqnSkeMF:-NufTn-vA5qXk4pZZBmi:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1634)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: period.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1641 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSZQb44xc9y-QHx21:-NufU0RS76yOJ1uGiDmj:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1641)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: period.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1642 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSoCaDUnpIslmIm58:-NufU1eg52VXUCFx-gUh:b-szepm7) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1642)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

We should never call native `sleep()` in tests. It's generally considered bad practice. In this case you probably want to hijack the time function inside the scheduler.
</blockquote></details>

I changed it so that it calls a function which sets the `last_update_timestamp` instead of sleeping.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1652 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSaK30p7ocNIAEjJV:-NufUGBX2YLq7zW_ajTh:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1652)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: period.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1665 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSdUSA6u9acwM1urv:-NufUHuI3JY4vRJDKJ5s:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1665)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: period.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1669 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSfmb9BKbbODj8vv8:-NufUImc4qBu9lrE1-zh:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1669)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: missing message.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adam-singer reviewed (commented):
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, Cargo Dev / macos-13, Remote / large-ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), macos-13, windows-2022 / stable, zig-cc ubuntu-20.04, zig-cc ubuntu-22.04, and 7 discussions need to be resolved



<!-- Sent from Reviewable.io -->

allada reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and 9 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwTBf-DF1Q5hn_pZEG:-NujA-SlCCF4GSMHUhCi:bkg771q) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

I changed it to be AtomicU64 and it works great
</blockquote></details>

Why `Arc`?
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 471 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwQbDM0-6DJULUOcga:-NujAsX823qQjkravtS7:b3zpf6w) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L471)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

`let queued_actions = self.queued_actions.iter().rev().cloned().collect()` gives a type error. I tried a bunch of different things to get it to work but couldn't figure it out.
</blockquote></details>

Did you instruct `.collect()` what the output container type was? Remember `.collect()` does not know what the type of the output will be, it just implements a trait that many containers implement. In this case you can use something like:
```rs
let queued_actions: Vec<(ActionInfo, Action)> = self.queued_actions.iter().rev().cloned().collect()
```
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 475 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWE3012nhQMnPFxd54:-NujBPuc25ZgoM78_X-h:btzztll) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L475)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

Do_try_match is the slowest part of the scheduler. If we capture it before the loop and it takes a substantial amount of time to complete, it might fail to time out some actions until the next pass.

Whether or not that it acceptable is something I'm not sure about
</blockquote></details>

I would not over optimize that. if `.do_try_match` is taking more than 1 second we are in serious trouble.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwT5As4BIfE0gOh9M_:-NujCDhG8UVUeQfgDUvx:b-fpsjam) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

I think we discussed this previously and decided that putting this in do_try_match was preferable to creating a new spawn which iterates to check and sleeps. Is there another method of checking when the client drops?
</blockquote></details>

Yeah, the cleanup happens here, but flagging when clients disconnect can/should happen when the event happens.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 60 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NujA7v64WZOMIAyoREC:-NujA7v64WZOMIAyoRED:b-7kooul) ([raw file](https://github.com/TraceMachina/nativelink/blob/2e287e5dba0cc8c19d5363e1b8fe350ed80000bc/nativelink-scheduler/src/simple_scheduler.rs#L60)):*
> ```XML
> const DEFAULT_MAX_JOB_RETRIES: usize = 3;
> 
> /// Default timeout for actions without any listeners
> ```

nit: period.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 363 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NujAORA6T-Ikyx5dfys:-NujAORA6T-Ikyx5dfyt:b-2yo8ud) ([raw file](https://github.com/TraceMachina/nativelink/blob/2e287e5dba0cc8c19d5363e1b8fe350ed80000bc/nativelink-scheduler/src/simple_scheduler.rs#L363)):*
> ```XML
>     }
> 
>     fn set_action_last_update_for_test(
> ```

nit: Inline this below, we want to limit the number of these floating around. These functions are generally allowed to be more ugly, since they are not supposed to be public facing.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 473 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NujBfEpDEscU1lAJMCF:-NujBfEpDEscU1lAJMCG:b-6c1ew7) ([raw file](https://github.com/TraceMachina/nativelink/blob/2e287e5dba0cc8c19d5363e1b8fe350ed80000bc/nativelink-scheduler/src/simple_scheduler.rs#L473)):*
> ```XML
>             // add update to queued action update timestamp here
>             let action = self.queued_actions.get_mut(&action_info).unwrap();
>             let now = SystemTime::now()
> ```

Using `SystemTime` inlined like this will make testing really difficult. I suggest making a constructor that you can pass in the time function.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1642 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSoCaDUnpIslmIm58:-NujCVTL65A8jGmSQb8j:bswpotf) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1642)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

I changed it so that it calls a function which sets the `last_update_timestamp` instead of sleeping.
</blockquote></details>

But we are back to the original problem since we are relying on wall time because we are now using `SystemTime`.


<!-- Sent from Reviewable.io -->

zbirenbaum reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and 9 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwTBf-DF1Q5hn_pZEG:-NujpXRtD2QdzyyHDzmQ:b-dvjgkw) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Why `Arc`?
</blockquote></details>

I wasn't sure where we landed on when to update this. I can take it out for now but it will need to be an Arc if we are updating this from another thread that detects when the user disconnects.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 471 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwQbDM0-6DJULUOcga:-NujrdUb3UWpaxhswX71:b-svbjke) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L471)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Did you instruct `.collect()` what the output container type was? Remember `.collect()` does not know what the type of the output will be, it just implements a trait that many containers implement. In this case you can use something like:
```rs
let queued_actions: Vec<(ActionInfo, Action)> = self.queued_actions.iter().rev().cloned().collect()
```
</blockquote></details>

I did, I tried all the following and more but none of them worked:
`let queued_actions: Vec<(ActionInfo, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(&ActionInfo, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(ActionInfo, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(&ActionInfo, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(Arc<ActionInfo>, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(Arc<ActionInfo>, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();
`let queued_actions: Vec<(&Arc<ActionInfo>, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(&Arc<ActionInfo>, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`

Here's an example of the full error:
 │     the method `collect` exists for struct `Cloned<Rev<Iter<'_, Arc<ActionInfo>, AwaitedAction>>>`, but its trait bounds were not satisfied rustc (E0599) [456, 106]
 │      the full type name has been written to '/Users/zach/Development/nativelink/target/debug/deps/nativelink_scheduler-1259087a177e1a06.long-type-1362048726443463511.txt' 
 │      the following trait bounds were not satisfied: 
 │      `<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>> as std::iter::Iterator>::Item = &_` 
 │      which is required by `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      which is required by `&mut std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │     the method `collect` exists for struct `Cloned<Rev<Iter<'_, Arc<ActionInfo>, AwaitedAction>>>`, but its trait bounds were not satisfied rustc (E0599) [456, 106]
 │      the full type name has been written to '/Users/zach/Development/nativelink/target/debug/deps/nativelink_scheduler-d0f178a39f61bd0a.long-type-2575323896278722321.txt' 
 │      the following trait bounds were not satisfied: 
 │      `<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>> as std::iter::Iterator>::Item = &_` 
 │      which is required by `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      which is required by `&mut std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │     the method call chain might not have had the expected associated types rustc (E0271) [456, 84]
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwT5As4BIfE0gOh9M_:-Nujqmtp0OXxWfZrf0Ua:b-ume55j) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Yeah, the cleanup happens here, but flagging when clients disconnect can/should happen when the event happens.
</blockquote></details>

How/where would we detect that the client disconnect has occurred? I was under the impression the only way to do this was to have a spawn which loops over the actions to check their receiver count.

I thought with disconnects it isn't possible to know until you request or try to send something to the other side
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1642 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSoCaDUnpIslmIm58:-NujuDg89_BiyA9D5T4C:b-6y7mom) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1642)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

But we are back to the original problem since we are relying on wall time because we are now using `SystemTime`.
</blockquote></details>

I'm a bit confused on a couple things here.

Where is the time function set/used in the scheduler? I see SystemTime used other places inside of it but not a `now_fn` like `RunningActionsManagerImpl` has.

I don't mind changing it to use that structure, but I just want to make sure that's something we want to include in this PR.

I'm also very confused about why we are relying on wall time here, the actual time doesn't matter at all unless I'm missing something huge.

The comparison done in the code to determine whether or not to time out is:
```
// If this is true the action is marked as old
running_action.get_last_update_timestamp() + self.disconnect_timeout_s < now
```
where:
`running_action.get_last_update_timestamp()` returns some uint
`now` is the current time since unix epoch in seconds
`self.disconnect_timeout_s` is a value like `1`

if we set `running_action.get_last_update_timestamp()` to be `SystemTime::now() - (self.disconnect_timeout_s+1)` then even if no time at all elapses between the first call to `SystemTime` in the test and the second one in `do_try_match` we end up with:

`now - (disconnect_timeout_s-1) + disconnect_timeout_s < now` = `-1 < 0` which will always return true

If the thought is that something like daylight savings or some system setting change could mess up the system time between the two calls to this test, we could just set the last_update_timestamp to 0 which would make it

`January 1st, 1970 + disconnect_timeout_s < now`


<!-- Sent from Reviewable.io -->

zbirenbaum reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Analyze (javascript-typescript), Bazel Dev / ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), pre-commit-checks, ubuntu-20.04 / stable, vale, and 8 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 363 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NujAORA6T-Ikyx5dfys:-Nuk-Cb-5CmI-hOUSaWN:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/2e287e5dba0cc8c19d5363e1b8fe350ed80000bc/nativelink-scheduler/src/simple_scheduler.rs#L363)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: Inline this below, we want to limit the number of these floating around. These functions are generally allowed to be more ugly, since they are not supposed to be public facing.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

allada reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and pending CI: Bazel Dev / ubuntu-22.04, asan / ubuntu-22.04, docker-compose-compiles-nativelink (20.04), docker-compose-compiles-nativelink (22.04), macos-13, ubuntu-22.04, zig-cc ubuntu-20.04, zig-cc ubuntu-22.04, and 8 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwTBf-DF1Q5hn_pZEG:-NukNmvx-PcpCYGA_zIs:bjb6wwy) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

I wasn't sure where we landed on when to update this. I can take it out for now but it will need to be an Arc if we are updating this from another thread that detects when the user disconnects.
</blockquote></details>

If we do that we will change the type. In this case we are not even sure how the future will end up, so keep it as simple as possible for now (no Arc).
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwT5As4BIfE0gOh9M_:-NukQ2pKAtJupjdickNn:b-6j35ub) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

How/where would we detect that the client disconnect has occurred? I was under the impression the only way to do this was to have a spawn which loops over the actions to check their receiver count.

I thought with disconnects it isn't possible to know until you request or try to send something to the other side
</blockquote></details>

Use `scope_guard` to detect when the stream is dropped in `execution_server::to_execute_stream()` that then calls a function on the scheduler to let the scheduler know that a client disconnected. The function would then take check if it's the last one subscribed and then the time.

Btw, this will make testing significantly easier with these public functions.


<!-- Sent from Reviewable.io -->

Commit:
Remove old actions with no listeners

Implement scheduler side removal of actions with no listeners. Adds
disconnect_timeout_s configuration field with default of 60s. If the
client waiting on a given action is disconnected for longer than this
duration without reconnecting the scheduler will stop tracking it. This
does not remove it from the worker if the job has already been
dispatched.

fixes TraceMachina#338

adam-singer reviewed (commented):
> Reviewed 10 of 11 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and 12 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 837 at r6](https://reviewable.io/reviews/TraceMachina/nativelink/778#-Nv5RkpqDoNP9iXYl_YF:-Nv5RkpqDoNP9iXYl_YG:b1xfh1h) ([raw file](https://github.com/TraceMachina/nativelink/blob/624eda2e29d31b33e2bf4989278d7528d0b1ef97/nativelink-scheduler/src/simple_scheduler.rs#L837)):*
> ```XML
> #[async_trait]
> impl ActionScheduler for SimpleScheduler {
>     fn notify_client_disconnected(&self, unique_qualifier: ActionInfoHashKey) {
> ```

When short returning in some of the `let/else` expressions, should we `debug`/`warn` log, it seems in some of the cases it might be useful to know that we are hitting what seems unexpected state or racing states?
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 837 at r6](https://reviewable.io/reviews/TraceMachina/nativelink/778#-Nv5SEZe9c4_cN5zO7xL:-Nv5SEZe9c4_cN5zO7xM:b-pswmea) ([raw file](https://github.com/TraceMachina/nativelink/blob/624eda2e29d31b33e2bf4989278d7528d0b1ef97/nativelink-scheduler/src/simple_scheduler.rs#L837)):*
> ```XML
> #[async_trait]
> impl ActionScheduler for SimpleScheduler {
>     fn notify_client_disconnected(&self, unique_qualifier: ActionInfoHashKey) {
> ```

We should be incrementing counters on different states here, not sure if a counter on `notify_client_disconnected` is useful (could be to match if we count successful removals from queues), we should counter on `inner.queued_actions*` removals and `inner.active_actions.remove`.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 859 at r6](https://reviewable.io/reviews/TraceMachina/nativelink/778#-Nv5RJ6SEMCfq8gjijMw:-Nv5RJ6SEMCfq8gjijMx:b-lqdy88) ([raw file](https://github.com/TraceMachina/nativelink/blob/624eda2e29d31b33e2bf4989278d7528d0b1ef97/nativelink-scheduler/src/simple_scheduler.rs#L859)):*
> ```XML
>         let weak_inner = Arc::downgrade(&self.inner);
> 
>         tokio::spawn(async move {
> ```

I think we should explain (comments) why we are spawning at this point and what the next stage of removing from queues means. 
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 899 at r6](https://reviewable.io/reviews/TraceMachina/nativelink/778#-Nv5R0TVEpoDDEIJlUt3:-Nv5R0TVEpoDDEIJlUt4:b3r6c82) ([raw file](https://github.com/TraceMachina/nativelink/blob/624eda2e29d31b33e2bf4989278d7528d0b1ef97/nativelink-scheduler/src/simple_scheduler.rs#L899)):*
> ```XML
>                         return;
>                     };
>                     // TODO: Send kill on worker signal - PR: #842.
> ```

Couldn't we just call `inner.active_actions.remove(&unique_qualifier);` since the match would fall through and return?


<!-- Sent from Reviewable.io -->

allada reviewed (commented):
Yes, I think `Instant` is the proper approach. It also solves a bit of other issues with importing `SysemTime`.

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778#-:-Nv7hPMxEnugiX0BExxE:be2sqxz)* status: 0 of 1 LGTMs obtained, and 12 discussions need to be resolved



<!-- Sent from Reviewable.io -->

zbirenbaum reviewed (commented):
> Dismissed @adam-singer and @allada from 4 discussions.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 0 of 1 LGTMs obtained, and 7 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 77 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwTBf-DF1Q5hn_pZEG:-Nv8YvJ5Cjc_Ds0cV1tu:b-4kirhb) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L77)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

If we do that we will change the type. In this case we are not even sure how the future will end up, so keep it as simple as possible for now (no Arc).
</blockquote></details>

Done. The var no longer exists so dismissing this.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 471 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwQbDM0-6DJULUOcga:-Nv8Z42S67A4JtalS-_m:bk5pxgz) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L471)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

I did, I tried all the following and more but none of them worked:
`let queued_actions: Vec<(ActionInfo, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(&ActionInfo, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(ActionInfo, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(&ActionInfo, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(Arc<ActionInfo>, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(Arc<ActionInfo>, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();
`let queued_actions: Vec<(&Arc<ActionInfo>, AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`
`let queued_actions: Vec<(&Arc<ActionInfo>, &AwaitedAction)> = self.queued_actions.iter().rev().cloned().collect();`

Here's an example of the full error:
 │     the method `collect` exists for struct `Cloned<Rev<Iter<'_, Arc<ActionInfo>, AwaitedAction>>>`, but its trait bounds were not satisfied rustc (E0599) [456, 106]
 │      the full type name has been written to '/Users/zach/Development/nativelink/target/debug/deps/nativelink_scheduler-1259087a177e1a06.long-type-1362048726443463511.txt' 
 │      the following trait bounds were not satisfied: 
 │      `<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>> as std::iter::Iterator>::Item = &_` 
 │      which is required by `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      which is required by `&mut std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │     the method `collect` exists for struct `Cloned<Rev<Iter<'_, Arc<ActionInfo>, AwaitedAction>>>`, but its trait bounds were not satisfied rustc (E0599) [456, 106]
 │      the full type name has been written to '/Users/zach/Development/nativelink/target/debug/deps/nativelink_scheduler-d0f178a39f61bd0a.long-type-2575323896278722321.txt' 
 │      the following trait bounds were not satisfied: 
 │      `<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>> as std::iter::Iterator>::Item = &_` 
 │      which is required by `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      `std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │      which is required by `&mut std::iter::Cloned<std::iter::Rev<std::collections::btree_map::Iter<'_, std::sync::Arc<nativelink_util::action_messages::ActionInfo>, simple_scheduler::AwaitedAction>>>: std::iter::Iterator` 
 │     the method call chain might not have had the expected associated types rustc (E0271) [456, 84]
</blockquote></details>

Done. `do_try_match` no longer has any edits and leaving this how it works on main so dismissing this.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 475 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWE3012nhQMnPFxd54:-Nv8ZfDV3eYgmeP06FY2:bk5pxgz) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L475)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

I would not over optimize that. if `.do_try_match` is taking more than 1 second we are in serious trouble.
</blockquote></details>

Done. `do_try_match` no longer has any edits and leaving this how it works on main so dismissing this.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 477 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwT5As4BIfE0gOh9M_:-NuzH7L55NmF-EBkrxu5:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L477)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Use `scope_guard` to detect when the stream is dropped in `execution_server::to_execute_stream()` that then calls a function on the scheduler to let the scheduler know that a client disconnected. The function would then take check if it's the last one subscribed and then the time.

Btw, this will make testing significantly easier with these public functions.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 548 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtWDJOt84qynHGY-3Hb:-Nv8_CsXCN7HGUnJK3PF:bk5pxgz) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/src/simple_scheduler.rs#L548)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done. `do_try_match` no longer has any edits and leaving this how it works on main so dismissing this.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 60 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NujA7v64WZOMIAyoREC:-Nv8YuH0ASTjLjX6Lrnd:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/2e287e5dba0cc8c19d5363e1b8fe350ed80000bc/nativelink-scheduler/src/simple_scheduler.rs#L60)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

nit: period.
</blockquote></details>

Done.
___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 473 at r4](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NujBfEpDEscU1lAJMCF:-Nv8_-VwDJBdw2Rn7Dwv:b-896fix) ([raw file](https://github.com/TraceMachina/nativelink/blob/2e287e5dba0cc8c19d5363e1b8fe350ed80000bc/nativelink-scheduler/src/simple_scheduler.rs#L473)):*
<details><summary><i>Previously, allada (Nathan (Blaise) Bruer) wrote…</i></summary><blockquote>

Using `SystemTime` inlined like this will make testing really difficult. I suggest making a constructor that you can pass in the time function.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

adam-singer reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778#-:-Nv9JHRB47RmjRt-evcJ:bnfp4nl)* status: 1 of 1 LGTMs obtained, and 4 discussions need to be resolved



<!-- Sent from Reviewable.io -->

allada reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 1 of 1 LGTMs obtained, and 3 discussions need to be resolved

___
*[`nativelink-scheduler/src/simple_scheduler.rs` line 485 at r7](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NvXZ2Lo7xvc8TxgHc0h:-NvXZ2Lo7xvc8TxgHc0i:b-jefw7h) ([raw file](https://github.com/TraceMachina/nativelink/blob/825aa7943a1705359a18e0bed5b96eeb7da2140b/nativelink-scheduler/src/simple_scheduler.rs#L485)):*
> ```XML
>             self.queued_actions.keys().rev().cloned().collect();
>         for action_info in action_infos {
>             // add update to queued action update timestamp here
> ```

don't forget this comment.
___
*[`nativelink-scheduler/tests/simple_scheduler_test.rs` line 1642 at r2](https://reviewable.io/reviews/TraceMachina/nativelink/778#-NtwSoCaDUnpIslmIm58:-NvXZNTyAmbtdXAvxKkK:bb4krlc) ([raw file](https://github.com/TraceMachina/nativelink/blob/ecadfb0ee7bb2062ad0023d8307572a3df44010f/nativelink-scheduler/tests/simple_scheduler_test.rs#L1642)):*
<details><summary><i>Previously, zbirenbaum (Zach Birenbaum) wrote…</i></summary><blockquote>

I'm a bit confused on a couple things here.

Where is the time function set/used in the scheduler? I see SystemTime used other places inside of it but not a `now_fn` like `RunningActionsManagerImpl` has.

I don't mind changing it to use that structure, but I just want to make sure that's something we want to include in this PR.

I'm also very confused about why we are relying on wall time here, the actual time doesn't matter at all unless I'm missing something huge.

The comparison done in the code to determine whether or not to time out is:
```
// If this is true the action is marked as old
running_action.get_last_update_timestamp() + self.disconnect_timeout_s < now
```
where:
`running_action.get_last_update_timestamp()` returns some uint
`now` is the current time since unix epoch in seconds
`self.disconnect_timeout_s` is a value like `1`

if we set `running_action.get_last_update_timestamp()` to be `SystemTime::now() - (self.disconnect_timeout_s+1)` then even if no time at all elapses between the first call to `SystemTime` in the test and the second one in `do_try_match` we end up with:

`now - (disconnect_timeout_s-1) + disconnect_timeout_s < now` = `-1 < 0` which will always return true

If the thought is that something like daylight savings or some system setting change could mess up the system time between the two calls to this test, we could just set the last_update_timestamp to 0 which would make it

`January 1st, 1970 + disconnect_timeout_s < now`
</blockquote></details>

If we move everything to `Instant` it will likely resolve all of these issues.


<!-- Sent from Reviewable.io -->

adam-singer reviewed (commented):
> Reviewed 6 of 8 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/TraceMachina/nativelink/778)* status: 1 of 1 LGTMs obtained, and 3 discussions need to be resolved



<!-- Sent from Reviewable.io -->

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/TraceMachina/nativelink?pullRequest=778) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/TraceMachina/nativelink?pullRequest=778) before we can accept your contribution.<br/><hr/>**Zach Birenbaum** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/TraceMachina/nativelink?pullRequest=778) it.</sub>
